File: 0206-reverse-linked-list/0206-reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        hey = []
        while head:
            hey.append(head.val)
            head = head.next

        if not hey:
            return None

        hey = hey[::-1]

        new_head = ListNode(hey[0])
        current = new_head
        for val in hey[1:]:
            current.next = ListNode(val)
            current = current.next

        return new_head

File: 0206-reverse-linked-list/0206-reverse-linked-list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        vector<int> values;

        // Store values from the linked list in a vector
        while (head) {
            values.push_back(head->val);
            head = head->next;
        }

        if (values.empty()) {
            return nullptr;
        }

        // Reverse the vector
        reverse(values.begin(), values.end());

        // Create new reversed linked list
        ListNode* new_head = new ListNode(values[0]);
        ListNode* current = new_head;
        for (int i = 1; i < values.size(); i++) {
            current->next = new ListNode(values[i]);
            current = current->next;
        }

        return new_head;
    }
};
